
import java.util.Stack;

public class Valid_Parenthesis {

    public static boolean validParenthesis(String str) {
        Stack<Character> s = new Stack<>();
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            if (ch == '(' || ch == '{' || ch == '[') {
                s.push(ch);
            } else {
                if (s.isEmpty()) {
                    return false;
                } else if ((s.peek() == '(' && ch == ')')
                        || (s.peek() == '{' && ch == '}')
                        || (s.peek() == '[' && ch == ']')) {
                    s.pop();
                }
            }

        }
        if (s.isEmpty()) {
            return true;
        }
        return false;

    }

    public static void main(String[] args) {
        String str = "({[]})]";
        System.out.println(validParenthesis(str));
    }
}
